### PR TITLE
[MLIR][DISC] Add RAL (Runtime abstraction layer) Dialect

### PR DESCRIPTION
**Pull request by wyzero**
DISC is a e2e flow, including both compiler side and runtime side. For
runtime side, we have different targeting environments (e.g. tensorflow,
pytorch, or sometimes even a standalone binary). In order to simplify
the design of the compiler side, we design a Runtime Abstraction Layer
(RAL) to sperate the compiler side and runtime side. Thus the compiler
side only need to target RAL itself and it is the responsibility of RAL
to handle the differences between different targeting environments.

One of the most important functions of RAL is to manage stateful
resources. To this end, it provides a context object, and hides all
stateful operations behind this context, thus the compiler side itself
doesn't need to care about the resource initialization. For example, a
kernel must be loaded before it can be launched on GPU. However, the
loading operation should only be taken once during the whole lifetime of
the context in order to achieve the best performance. Based on the
initialization-free interfaces provided by RAL, compiler side can focus
on its core optimization logic and lets the RAL to manage the resource
status.

The context mentioned above is passed as a parameter to the entry
function and all RAL APIs should always use the context as their first
argument. This CR also provides a pass to help to ensure this property.
The pass rewrites the entry function to make sure their first argument
is the context. For entry function, the pass also rewrites its inputs
and outputs. To be concrete, all the original inputs and outputs of the
entry function are received from and sent to RAL through a sequence of
RAL API calls correspondingly. The motivation behind this is to hide the
implementation details of I/Os. This design may also potentially enable
partial execution of the compiled module when some of the inputs are
ready.

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/tests/disc_ral_inject_execution_context.mlir [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// RUN: mlir-hlo-opt -disc-ral-inject-execution-context=entry-func-name=test \
+// RUN:  -canonicalize -lhlo-legalize-tensor-load-op %s -o - | FileCheck %s
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/tests/disc_ral_inject_execution_context.mlir:2]: Do you need the `-canonicalize -lhlo-legalize-tensor-load-op` part here?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/init.cc [hunk 2]
```diff
@@ -22,6 +22,7 @@ limitations under the License.
 void mlir::mhlo::registerAllMhloDialects(mlir::DialectRegistry &registry) {
   // clang-format off
   registry.insert<mlir::chlo::HloClientDialect,
+                  mlir::disc_ral::RalDialect,
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/init.cc:25]: I'm not convinced this should be registered here (but  I'm also not sure that lmhlo_gpu should be there as well).
Can you leave this out?

**Review by joker-eph — COMMENTED** [review]
LG overall!

**Review by joker-eph — COMMENTED** [review]
Can you add the missing CMakeLists.txt so that this all test and run in https://github.com/tensorflow/mlir-hlo ?

File: tensorflow/compiler/mlir/hlo/tests/disc_ral_inject_execution_context.mlir (hunk 1 above)
Comment (wyzero) [reply tensorflow/compiler/mlir/hlo/tests/disc_ral_inject_execution_context.mlir]: removed `-lhlo-legalize-tensor-load-op`

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/init.cc (hunk 2 above)
Comment (wyzero) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/init.cc]: Done!

**Comment by wyzero** [comment]
> Can you add the missing CMakeLists.txt so that this all test and run in [tensorflow/mlir-hlo](https://github.com/tensorflow/mlir-hlo) ?

Done

**Review by joker-eph — APPROVED** [review]

**Review by joker-eph — APPROVED** [review]

**Comment by wyzero** [comment]
@joker-eph Hi, it seems that importing to inside code review stage is failed. Does it need to re-run again?

**Comment by rthadur** [comment]
Seems auto-merge is not happening but the changes are now committed so we can close this. Thank you for the PR.

**Pull request closed without merging**